### PR TITLE
Extend PropagateInit interface.

### PR DESCRIPTION
**Pull request by rkaminsk**
This pull request extends the PropagateInit class to add constraints and perform propagation during propagator initialization.

### CODE REVIEW

**Comment by rkaminsk** [comment]
@BenKaufmann can you have a look at 732231d to check if this makes sense? I was especially unsure if it is okay to call propagate on the master solver during propagator initializtion like this.

**Comment by rkaminsk** [comment]
@BenKaufmann can you please also have a look at 713283f? I added a method to add weight constraints during propagator initialization.

**Comment by rkaminsk** [comment]
@BenKaufmann and there is a final one for adding literals to minimize in PropagateInit. Please have a look at ec8a6f9, too.

File: libclingo/clingo.h [hunk 1]
```diff
@@ -1088,23 +1094,62 @@ CLINGO_VISIBILITY_DEFAULT clingo_propagator_check_mode_t clingo_propagate_init_g
...
 //!
-//! @note The result is always set to true and might be removed in a future version of the API.
-//!
+//! @attention No further calls on the init object or functions on the assignment should be called when the result of this method is false.
 //!
 //! @param[in] init the target
 //! @param[in] clause the clause to add
 //! @param[in] size the size of the clause
-//! @param[out] result always true
+//! @param[out] result result indicating whether whether the problem became unsatisfiable
```
Comment (BenKaufmann) [inline libclingo/clingo.h]: There is a second `whether` that can be dropped

File: libclingo/src/clingocontrol.cc [hunk 2]
```diff
@@ -392,12 +393,46 @@ class ClingoPropagateInit : public PropagateInit {
...
+        Clasp::Var m = 0;
+        for (auto &x : lits) {
+            auto y = Clasp::decodeLit(x.lit);
+            m = std::max(m, y.var());
+            claspLits.push_back({y, x.weight});
+        }
+        if (m > 0 && !master.validVar(m)) {
+            master.acquireProblemVars();
+        }
+        return Clasp::WeightConstraint::create(*ctx.master(), Clasp::decodeLit(lit), claspLits, bound, eq ? Clasp::WeightConstraint::create_eq_bound : 0).ok();
```
Comment (BenKaufmann) [inline libclingo/src/clingocontrol.cc]: This looks ok to me but maybe we should consider moving the "acquireProblemVars()" logic to clasp to make it more consistent with clause creation.

File: libclingo/src/clingocontrol.cc [hunk 3]
```diff
@@ -392,12 +393,46 @@ class ClingoPropagateInit : public PropagateInit {
...
+    }
+    void addMinimize(Potassco::Lit_t literal, Potassco::Weight_t weight, Potassco::Weight_t priority) override {
+        auto &ctx = static_cast<Clasp::ClaspFacade*>(c_.claspFacade())->ctx;
+        auto &master = *ctx.master();
+        auto lit = Clasp::decodeLit(literal);
+        if (!master.validVar(lit.var())) { master.acquireProblemVars(); }
```
Comment (BenKaufmann) [inline libclingo/src/clingocontrol.cc]: See above

File: libclingo/src/clingocontrol.cc [hunk 4]
```diff
@@ -392,12 +393,46 @@ class ClingoPropagateInit : public PropagateInit {
...
+        ctx.addMinimize({lit, weight}, priority);
+    }
+    bool propagate() override {
+        auto &ctx = static_cast<Clasp::ClaspFacade*>(c_.claspFacade())->ctx;
+        if (ctx.master()->hasConflict()) { return false; }
+        return ctx.master()->propagate();
```
Comment (BenKaufmann) [inline libclingo/src/clingocontrol.cc]: Looks good to me - with the caveat regarding Sat-Preprocessing, propagating during init should be ok.

**Comment by rkaminsk** [comment]
Awesome, thanks. I'll address the Doxygen comments and also add Lua and Python bindings next. Would be nice if you could take care of the `acquireProblemVars` logic at some point. I guess it's not a big deal anyway.

**Comment by BenKaufmann** [comment]
> Would be nice if you could take care of the `acquireProblemVars` logic at some point. I guess it's not a big deal anyway.
> 

I added the necessary changes in https://github.com/potassco/clasp/commit/13ae0ea46ae79e7b8359825a68c2f5ef46f637af

**Comment by rkaminsk** [comment]
> > Would be nice if you could take care of the `acquireProblemVars` logic at some point. I guess it's not a big deal anyway.
> 
> I added the necessary changes in [potassco/clasp@13ae0ea](https://github.com/potassco/clasp/commit/13ae0ea46ae79e7b8359825a68c2f5ef46f637af)

Thanks, using it now.

**Comment by MaxOstrowski** [comment]
Looks awesome. Thanks a lot.

**Pull request merged**